### ISSUE TITLE
[Feature Request] Failover on connection refused

### ISSUE THREAD

**Issue by SilentHunter44** [grobian/carbon-c-relay#276]
Hi,

I’m currently trying to implement a high availability metrology stack between multiple datacenters using multiple carbon-c-relay instances on the same machines, to feed 2 Carbon/Whisper backends in parallel with the same data.

Roughly, we are planning on the following infrastructure:
![4hm2un2v](https://cloud.githubusercontent.com/assets/5552420/26212689/9ba6714a-3bf6-11e7-8c87-3095cbb7cb1b.png)

The main idea is to forward all the received data to the 2 Carbon/Whisper backends (first layer of relays on the schema, the forward ones).

If one of the Carbon/Whisper backend nodes go offline, we write its metrics into a logfile to avoid queue overflows and to reinject them later (second and third layer of relays).

The failover relays are aimed to perform a failover routing between Carbon/Whisper backend and the file relays, which just write all input metrics into a logfile.

The problem we are facing is that when we perform a maintenance on a backend node (by stopping Carbon but not the entire host), the failover relays don’t switch to the other destination, quickly ending in a queue overflow case. In the failover relays logs, I can see the following lines:
```
[2017-05-18 14:24:09] (ERR) failed to connect() for <BACKEND_IP>:2003: Connection refused
[...]
[2017-05-18 14:38:56] (ERR) server <BACKEND_IP>:2003: OK after probe
```

So if I understand well, the relay detects that the connection is no longer possible, but does not switch to the other destination, until the connection can be established again.

Would it be possible to implement a kind of failover switch condition (when the connection is refused by the destination but the host itself is still "alive")?

Thank you in advance for your help !

PS : We are using the following configurations on our relays:

- forward
```sh
cluster next-hop
    forward
        127.0.0.1:2103 # failover_1
        127.0.0.1:2203 # failover_2
    ;
match *
    send to next-hop
    stop
    ;
```

- failover_1
```sh
cluster next-hop
    failover
        <CARBON_1_IP>:2003
        127.0.0.1:2113 # file_1
    ;
match *
    send to next-hop
    stop
    ;
```

- failover_2
```sh
cluster next-hop
    failover
        <CARBON_2_IP>:2003
        127.0.0.1:2213 # file_2
    ;
match *
    send to next-hop
    stop
    ;
```

- file
```sh
cluster logfile
    file
        <DESTINATION_FILE>
    ;
match *
    send to logfile
    stop
    ;
```

**Comment by grobian** [comment]
In the current implementation, the next server in the list should be used as soon as failure is detected. A delay of ~1.5 seconds is used before the pending metrics of the failed server are passed onto the next.

**Comment by grobian** [comment]
your setup looks similar to what I setup at Booking, which is much like this:

```
cluster foo_dc1 fnv1a_ch
  ip11 ip12 ip13 ..
;
cluster foo_dc2 fnv1a_ch
  ip21 ip22 ip23 ...
;

match ^prefix_of_foo
  send to foo_dc1 foo_dc2
  stop;
```
This config would run in both dc1 and dc2 on a number of dedicated nodes with queue-sizes setup to deal with the load and allow for queuing up until a reasonable memory pressure, etc.

Since the objective of carbon-c-relay was to eliminate the need for multiple relays to be stacked in order to just route to multiple clusters, etc. I'm confident your config can do with less routers.  At least failover_1 and failover_2 can be folded into forward.  Overflowing to the disk is much like issue #271, where the relay would transparently buffer metrics on disk to be re-read when resources allow.  While I understand the need, the original design idea was to always be in memory, in other words to get rid of metrics ASAP -- they need to end up on their final destination, not get stuck in routing somewhere.

**Comment by SilentHunter44** [comment]
I have merged the forward relays and the failover ones into a unique input relay as you said (by drawing inspiration from your configuration example) and.. It works like a charm now!

Thank you very much for your time and for those great advices!

**Comment by grobian** [comment]
Closing this issue.  If you need more, reopen or open a new one.  Thanks!
